README.md
## NB_Generic

[![arduino-library-badge](https://www.ardu-badge.com/badge/NB_Generic.svg?)](https://www.ardu-badge.com/NB_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/NB_Generic.svg)](https://github.com/khoih-prog/NB_Generic/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/NB_Generic/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/NB_Generic.svg)](http://github.com/khoih-prog/NB_Generic/issues)

---
---

## Table of Contents


* [Why do we need this NB_Generic library](#why-do-we-need-this-NB_Generic-library)
  * [Features](#features)
  * [Supported modems](#supported-modems)
  * [To be Supported modems](#to-be-supported-modems)
  * [Currently supported Boards](#currently-supported-boards)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Packages' Patches](#packages-patches)
  * [1. For Adafruit nRF52840 and nRF52832 boards](#1-for-adafruit-nRF52840-and-nRF52832-boards)
  * [2. For Teensy boards](#2-for-teensy-boards)
  * [3. For Arduino SAM DUE boards](#3-for-arduino-sam-due-boards)
  * [4. For Arduino SAMD boards](#4-for-arduino-samd-boards)
      * [For core version v1.8.10+](#for-core-version-v1810)
      * [For core version v1.8.9-](#for-core-version-v189-)
  * [5. For Adafruit SAMD boards](#5-for-adafruit-samd-boards)
  * [6. For Seeeduino SAMD boards](#6-for-seeeduino-samd-boards)
  * [7. For STM32 boards](#7-for-stm32-boards) 
    * [7.1. For STM32 boards to use LAN8720](#71-for-stm32-boards-to-use-lan8720)
    * [7.2. For STM32 boards to use Serial1](#72-for-stm32-boards-to-use-serial1)
  * [8. For RP2040-based boards using Earle Philhower arduino-pico core](#8-for-rp2040-based-boards-using-earle-philhower-arduino-pico-core)
    * [8.1. To use BOARD_NAME](#81-to-use-board_name)
    * [8.2. To avoid compile error relating to microsecondsToClockCycles](#82-to-avoid-compile-error-relating-to-microsecondstoclockcycles)
  * [9. For Portenta_H7 boards using Arduino IDE in Linux](#9-for-portenta_h7-boards-using-arduino-ide-in-linux)
* [Libraries' Patches](#libraries-patches)
  * [1. For Adafruit_MQTT_Library](#1-for-adafruit_mqtt_library)
* [HOWTO Install esp32 core for ESP32-S2 (Saola, AI-Thinker ESP-12K) and ESP32-C3 boards into Arduino IDE](#howto-install-esp32-core-for-esp32-s2-saola-ai-thinker-esp-12k-and-esp32-c3-boards-into-arduino-ide)
  * [1. Save the original esp32 core](#1-save-the-original-esp32-core)
  * [2. Install esp32 core v1.0.6](#2-install-esp32-core-v106)
    * [2.1 Install esp32 core](#21-install-esp32-core)
    * [2.2 Download latest zip with esp32-s2 support](#22-download-latest-zip-with-esp32-s2-support)
    * [2.3 Unzip](#23-unzip)
    * [2.3 Update esp32 core directories](#24-update-esp32-core-directories)
  * [3. Download tools for ESP32-S2](#3-download-tools-for-esp32-s2) 
    * [3.1 Download Toolchain for Xtensa (ESP32-S2) based on GCC](#31-download-toolchain-for-xtensa-esp32-s2-based-on-gcc)
    * [3.2 Download esptool](#32-download-esptool)
    * [3.3 Unzip](#33-unzip)
  * [4. Update tools](#4-update-tools)
    * [4.1 Update Toolchain](#41-update-toolchain)
    * [4.2 Update esptool](#42-update-esptool)
  * [5. Download tools for ESP32-C3](#5-download-tools-for-esp32-c3)
  * [6. esp32-s2 WebServer Library Patch](#6-esp32-s2-webserver-library-patch)
* [Note for Platform IO using ESP32 LittleFS](#note-for-platform-io-using-esp32-littlefs)
* [Reference](#reference)
* [Documentation](#documentation)
  * [1. How it works](#1-how-it-works)
  * [2. Library structure](#2-library-structure)
  * [3. Library main classes](#3-library-main-classes)
    * [NB Class](#nb-class)
    * [NB_SMS Class](#nb_sms-class)
    * [GPRS Class](#gprs-class)
    * [NBClient Class](#nbclient-class)
    * [NBSSLClient Class](#nbsslclient-class)
    * [NBModem Class](#nbmodem-class)
    * [NBScanner Class](#nbscanner-class)
    * [NBPIN Class](#nbpin-class)
    * [NBUDP Class](#nbudp-class)
* [Configuration Notes](#configuration-notes)
  * [1. How to select which NB-IoT/LTE-M/GPRS module to use](#1-how-to-select-which-nb-iotlte-mgprs-module-to-use)
  * [2. How to select which Serial port to communicate to NB module](#2-how-to-select-which-serial-port-to-communicate-to-nb-module)
  * [3. How to select extra pins to control NB module](#3-how-to-select-extra-pins-to-control-nb-module)
* [Examples](#examples)
  * [ 1. GPRSUdpNtpClient](examples/GPRSUdpNtpClient)
  * [ 2. NBSSLWebClient](examples/NBSSLWebClient)
  * [ 3. NBWebClient](examples/NBWebClient)
  * [ 4. ReceiveSMS](examples/ReceiveSMS)
  * [ 5. SendSMS](examples/SendSMS)
* [Tools](#tools)
  * [ 6. ChooseRadioAccessTechnology](examples/Tools/ChooseRadioAccessTechnology)
  * [ 7. NBScanNetworks](examples/Tools/NBScanNetworks)
  * [ 8. PinManagement](examples/Tools/PinManagement)
  * [ 9. SerialNBPassthrough](examples/Tools/SerialNBPassthrough)
  * [10. TestGPRS](examples/Tools/TestGPRS)
  * [11. TestModem](examples/Tools/TestModem)
  * [Advanced Examples](#advanced-examples)
    * [MQTT](#mqtt)
      * [26. NB_MQTT](examples/Advanced/NB_MQTT)
      * [27. NB_MQTTClient_Auth](examples/Advanced/NB_MQTTClient_Auth)
      * [28. NB_MQTTClient_Basic](examples/Advanced/NB_MQTTClient_Basic)
      * [29. NB_MQTT_Medium](examples/Advanced/NB_MQTT_Medium)
      * [30. NB_MQTT_ThingStream](examples/Advanced/NB_MQTT_ThingStream)
    * [Blynk](#blynk)
      * [31. NB_Blynk](examples/Advanced/NB_Blynk)

changelog.md
## NB_Generic

[![arduino-library-badge](https://www.ardu-badge.com/badge/NB_Generic.svg?)](https://www.ardu-badge.com/NB_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/NB_Generic.svg)](https://github.com/khoih-prog/NB_Generic/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/NB_Generic/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/NB_Generic.svg)](http://github.com/khoih-prog/NB_Generic/issues)

---
---

## Table of Contents


* [Changelog](#changelog)
  * [Release v1.3.1](#release-v131)
  * [Release v1.3.0](#release-v130)
  * [Release v1.2.0](#release-v120)
  * [Major Release v1.1.0](#major-release-v110)
  * [Release v1.0.1](#release-v101)
  * [Release v1.0.0](#release-v100)


---
---

## Changelog

### Release v1.3.1

1. Update `Packages' Patches`

### Release v1.3.0

1. Add support to Adafruit nRF52 core v0.22.0+
2. Add support to Raytac MDBT50Q_RX Dongle

### Release v1.2.0

1. Add support to RP2040-based boards, such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, using [**Earle Philhower's arduino-pico core**](https://github.com/earlephilhower/arduino-pico).
2. Add support to RP2040-based boards, such as **NANO_RP2040_CONNECT, RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed).

### Major Release v1.1.0

1. Rewrite to isolate AT-command hardware-specific code to prepare for supporting more NB-IoT/LTE-M/GPRS modems
2. Add FileUtils-related examples
3. Make `NB_RESETN`, `NB_PWR` optional for u-blox NB-IoT/LTE-M/GPRS modems
4. Permit configure baudrate of NB-IoT/LTE-M/GPRS Modem
5. Reduce hardcoding
6. Add `Bug Report` and `Feature Request` templates

### Release v1.0.1

1. Add Advanced examples (MQTT, Blynk)

### Release v1.0.0

1. Initial coding to add support to many boards besides MKRNB 1500 / u-blox NB-IoT/LTE-M/GPRS modems (Sara R4xx NB-IoT/LTE-M modules (SARA-R404M, SARA-R410M, SARA-R412M, SARA-N410, etc)





src/NB_Boards_Generic.h
/*********************************************************************************************************************************
  NB_Boards_Generic.h
  
  For ESP8266, ESP32, SAMD21/SAMD51, nRF52, SAM DUE, Teensy and STM32 with NB modules
  
  NB_Generic is a library for the ESP8266, ESP32, SAMD21/SAMD51, nRF52, SAM DUE, Teensy and STM32 with NB modules
  
  Based on and modified from MKRNB Library (https://github.com/arduino-libraries/MKRNB)
  
  Built by Khoi Hoang https://github.com/khoih-prog/NB_Generic
  Licensed under GNU Lesser General Public License
  
  Copyright (C) 2018  Arduino AG (http://www.arduino.cc/)
  This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License 
  as published by the Free Software Foundation, either version 2.1 of the License, or (at your option) any later version.
  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
  of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
  You should have received a copy of the GNU General Public License along with this program.
  If not, see <https://www.gnu.org/licenses/>.  
 
  Version: 1.3.1
  
  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0    K Hoang     18/03/2021 Initial public release to add support to many boards / modules besides MKRNB 1500 / SARA R4
  1.0.1    K Hoang     18/03/2021 Add Advanced examples (MQTT, Blynk)
  1.1.0    K Hoang     19/03/2021 Rewrite to prepare for supporting more GSM/GPRS modules. Add FileUtils examples.
  1.2.0    K Hoang     28/06/2021 Add support to RP2040-based boards using Arduino mbed or Arduino-pico core
  1.3.0    K Hoang     14/08/2021 Add support to Adafruit nRF52 core v0.22.0+
  1.3.1    K Hoang     10/10/2021 Update `Packages' Patches`
 **********************************************************************************************************************************/

#pragma once

#ifndef _NB_BOARDS_GENERIC_H_INCLUDED
#define _NB_BOARDS_GENERIC_H_INCLUDED

#include <Arduino.h>

#if    ( defined(ARDUINO_SAMD_ZERO) || defined(ARDUINO_SAMD_MKR1000) || defined(ARDUINO_SAMD_MKRWIFI1010) \
      || defined(ARDUINO_SAMD_NANO_33_IOT) || defined(ARDUINO_SAMD_MKRFox1200) || defined(ARDUINO_SAMD_MKRWAN1300) || defined(ARDUINO_SAMD_MKRWAN1310) \
      || defined(ARDUINO_SAMD_MKRGSM1400) || defined(ARDUINO_SAMD_MKRNB1500) || defined(ARDUINO_SAMD_MKRVIDOR4000) || defined(__SAMD21G18A__) \
      || defined(ARDUINO_SAMD_CIRCUITPLAYGROUND_EXPRESS) || defined(__SAMD21E18A__) || defined(__SAMD51__) || defined(__SAMD51J20A__) || defined(__SAMD51J19A__) \
      || defined(__SAMD51G19A__) || defined(__SAMD51P19A__) || defined(__SAMD21G18A__) )
  #if defined(NB_USE_SAMD)
    #undef NB_USE_SAMD
  #endif
  #define NB_USE_SAMD      true

#elif (defined(NRF52840_FEATHER) || defined(NRF52832_FEATHER) || defined(NRF52_SERIES) || defined(ARDUINO_NRF52_ADAFRUIT) || \
       defined(NRF52840_FEATHER_SENSE) || defined(NRF52840_ITSYBITSY) || defined(NRF52840_CIRCUITPLAY) || \
       defined(NRF52840_CLUE) || defined(NRF52840_METRO) || defined(NRF52840_PCA10056) || defined(PARTICLE_XENON) || \
       defined(MDBT50Q_RX) || defined(NINA_B302_ublox) || defined(NINA_B112_ublox) )
  #if defined(NB_USE_NRF528XX)
    #undef NB_USE_NRF528XX
  #endif
  #define NB_USE_NRF528XX      true
  
  #include <Adafruit_TinyUSB.h>
 
#elif ( defined(ARDUINO_SAM_DUE) || defined(__SAM3X8E__) )
  #if defined(NB_USE_SAM_DUE)
    #undef NB_USE_SAM_DUE
  #endif
  #define NB_USE_SAM_DUE      true

#elif ( defined(STM32F0) || defined(STM32F1) || defined(STM32F2) || defined(STM32F3)  ||defined(STM32F4) || defined(STM32F7) || \
       defined(STM32L0) || defined(STM32L1) || defined(STM32L4) || defined(STM32H7)  ||defined(STM32G0) || defined(STM32G4) || \
       defined(STM32WB) || defined(STM32MP1) )
  #if defined(NB_USE_STM32)
    #undef NB_USE_STM32
  #endif
  #define NB_USE_STM32      true

#elif ( defined(ARDUINO_NANO_RP2040_CONNECT) || defined(ARDUINO_ARCH_RP2040) || defined(ARDUINO_RASPBERRY_PI_PICO) || \
      defined(ARDUINO_GENERIC_RP2040) || defined(ARDUINO_ADAFRUIT_FEATHER_RP2040) )
  #if defined(NB_USE_RP2040)
    #undef NB_USE_RP2040
  #endif
  #define NB_USE_RP2040      true
    
#endif

///////////////////////////////////////////

#if defined(NB_USE_SAMD)

  #if !defined(ARDUINO_SAMD_MKRNB1500)
    // Must define these pins for SAMD
    #if !defined(SerialNB)
      #warning Using default SerialNB = Serial1

      #define SerialNB   Serial1
    #else
      #warning Using SerialNB from sketch  
    #endif
    
    #if !( defined(NB_RESETN) && defined(NB_PWR) )
      #warning Using default NB_RESETN(10), NB_PWR(11)
      #define NB_RESETN  (10u)
      #define NB_PWR     (11u)

    #else
      #warning Using NB_RESETN and NB_PWR pins from sketch  
    #endif
    
    #warning Using SAMD different from MKRNB1500
  #endif

  #if ( defined(ARDUINO_SAMD_ZERO) && !defined(SEEED_XIAO_M0) )
    #define BOARD_TYPE      "SAMD Zero"
  #elif defined(ARDUINO_SAMD_MKR1000)
    #define BOARD_TYPE      "SAMD MKR1000"
  #elif defined(ARDUINO_SAMD_MKRWIFI1010)
    #define BOARD_TYPE      "SAMD MKRWIFI1010"
  #elif defined(ARDUINO_SAMD_NANO_33_IOT)
    #define BOARD_TYPE      "SAMD NANO_33_IOT"
  #elif defined(ARDUINO_SAMD_MKRFox1200)
    #define BOARD_TYPE      "SAMD MKRFox1200"
  #elif ( defined(ARDUINO_SAMD_MKRWAN1300) || defined(ARDUINO_SAMD_MKRWAN1310) )
    #define BOARD_TYPE      "SAMD MKRWAN13X0"
  #elif defined(ARDUINO_SAMD_MKRGSM1400)
    #define BOARD_TYPE      "SAMD MKRGSM1400"
  #elif defined(ARDUINO_SAMD_MKRNB1500)
    #define BOARD_TYPE      "SAMD MKRNB1500"
  #elif defined(ARDUINO_SAMD_MKRVIDOR4000)
    #define BOARD_TYPE      "SAMD MKRVIDOR4000"
  #elif defined(ARDUINO_SAMD_CIRCUITPLAYGROUND_EXPRESS)
    #define BOARD_TYPE      "SAMD ARDUINO_SAMD_CIRCUITPLAYGROUND_EXPRESS"
  #elif defined(ADAFRUIT_FEATHER_M0_EXPRESS)
    #define BOARD_TYPE      "SAMD21 ADAFRUIT_FEATHER_M0_EXPRESS"
  #elif defined(ADAFRUIT_METRO_M0_EXPRESS)
    #define BOARD_TYPE      "SAMD21 ADAFRUIT_METRO_M0_EXPRESS"
  #elif defined(ADAFRUIT_CIRCUITPLAYGROUND_M0)
    #define BOARD_TYPE      "SAMD21 ADAFRUIT_CIRCUITPLAYGROUND_M0"
  #elif defined(ADAFRUIT_GEMMA_M0)
    #define BOARD_TYPE      "SAMD21 ADAFRUIT_GEMMA_M0"
  #elif defined(ADAFRUIT_TRINKET_M0)
    #define BOARD_TYPE      "SAMD21 ADAFRUIT_TRINKET_M0"
  #elif defined(ADAFRUIT_ITSYBITSY_M0)
    #define BOARD_TYPE      "SAMD21 ADAFRUIT_ITSYBITSY_M0"
  #elif defined(ARDUINO_SAMD_HALLOWING_M0)
    #define BOARD_TYPE      "SAMD21 ARDUINO_SAMD_HALLOWING_M0"
  #elif defined(ADAFRUIT_METRO_M4_EXPRESS)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_METRO_M4_EXPRESS"
  #elif defined(ADAFRUIT_GRAND_CENTRAL_M4)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_GRAND_CENTRAL_M4"
  #elif defined(ADAFRUIT_FEATHER_M4_EXPRESS)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_FEATHER_M4_EXPRESS"
  #elif defined(ADAFRUIT_ITSYBITSY_M4_EXPRESS)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_ITSYBITSY_M4_EXPRESS"
    #define USE_THIS_SS_PIN       10
  #elif defined(ADAFRUIT_TRELLIS_M4_EXPRESS)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_TRELLIS_M4_EXPRESS"
  #elif defined(ADAFRUIT_PYPORTAL)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_PYPORTAL"
  #elif defined(ADAFRUIT_PYPORTAL_M4_TITANO)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_PYPORTAL_M4_TITANO"
  #elif defined(ADAFRUIT_PYBADGE_M4_EXPRESS)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_PYBADGE_M4_EXPRESS"
  #elif defined(ADAFRUIT_METRO_M4_AIRLIFT_LITE)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_METRO_M4_AIRLIFT_LITE"
  #elif defined(ADAFRUIT_PYGAMER_M4_EXPRESS)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_PYGAMER_M4_EXPRESS"
  #elif defined(ADAFRUIT_PYGAMER_ADVANCE_M4_EXPRESS)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_PYGAMER_ADVANCE_M4_EXPRESS"
  #elif defined(ADAFRUIT_PYBADGE_AIRLIFT_M4)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_PYBADGE_AIRLIFT_M4"
  #elif defined(ADAFRUIT_MONSTER_M4SK_EXPRESS)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_MONSTER_M4SK_EXPRESS"
  #elif defined(ADAFRUIT_HALLOWING_M4_EXPRESS)
    #define BOARD_TYPE      "SAMD51 ADAFRUIT_HALLOWING_M4_EXPRESS"
  #elif defined(SEEED_WIO_TERMINAL)
    #define BOARD_TYPE      "SAMD SEEED_WIO_TERMINAL"
  #elif defined(SEEED_FEMTO_M0)
    #define BOARD_TYPE      "SAMD SEEED_FEMTO_M0"
  #elif defined(SEEED_XIAO_M0)
    #define BOARD_TYPE      "SAMD SEEED_XIAO_M0"
  #elif defined(Wio_Lite_MG126)
    #define BOARD_TYPE      "SAMD SEEED Wio_Lite_MG126"
  #elif defined(WIO_GPS_BOARD)
    #define BOARD_TYPE      "SAMD SEEED WIO_GPS_BOARD"
  #elif defined(SEEEDUINO_ZERO)
    #define BOARD_TYPE      "SAMD SEEEDUINO_ZERO"
  #elif defined(SEEEDUINO_LORAWAN)
    #define BOARD_TYPE      "SAMD SEEEDUINO_LORAWAN"
  #elif defined(SEEED_GROVE_UI_WIRELESS)
    #define BOARD_TYPE      "SAMD SEEED_GROVE_UI_WIRELESS"
  #elif defined(__SAMD21E18A__)
    #define BOARD_TYPE      "SAMD21E18A"
  #elif defined(__SAMD21G18A__)
    #define BOARD_TYPE      "SAMD21G18A"
  #elif defined(__SAMD51G19A__)
    #define BOARD_TYPE      "SAMD51G19A"
  #elif defined(__SAMD51J19A__)
    #define BOARD_TYPE      "SAMD51J19A"
  #elif defined(__SAMD51J20A__)
    #define BOARD_TYPE      "SAMD51J20A"
  #elif defined(__SAM3X8E__)
    #define BOARD_TYPE      "SAM3X8E"
  #elif defined(__CPU_ARC__)
    #define BOARD_TYPE      "CPU_ARC"
  #elif defined(__SAMD51__)
    #define BOARD_TYPE      "SAMD51"
  #else
    #define BOARD_TYPE      "SAMD Unknown"
  #endif

#elif (NB_USE_SAM_DUE)

  // Must define these pins for SAM DUE
  #if !defined(SerialNB)
    #warning Using default SerialNB = Serial1

    #define SerialNB   Serial1
  #else
    #warning Using SerialNB from sketch  
  #endif
  
  #if !( defined(NB_RESETN) && defined(NB_PWR) )
    #warning Using default NB_RESETN(10), NB_PWR(11)
    #define NB_RESETN  (10u)
    #define NB_PWR     (11u)

  #else
    #warning Using NB_RESETN and NB_PWR pins from sketch  
  #endif
  
  #warning Using SAM DUE
 
  #define BOARD_TYPE      "SAM DUE"

#elif (NB_USE_NRF528XX)

  // Must define these pins for nRF52
  #if !defined(SerialNB)
    #warning Using default SerialNB = Serial1

    #define SerialNB   Serial1
  #else
    #warning Using SerialNB from sketch  
  #endif
  
  #if !( defined(NB_RESETN) && defined(NB_PWR) )
    #warning Using default NB_RESETN(10), NB_PWR(11)
    #define NB_RESETN  (10u)
    #define NB_PWR     (11u)

  #else
    #warning Using NB_RESETN and NB_PWR pins from sketch  
  #endif
  
  #warning Using nRF52

  #if defined(NRF52840_FEATHER)
    #define BOARD_TYPE      "NRF52840_FEATHER"
  #elif defined(NRF52832_FEATHER)
    #define BOARD_TYPE      "NRF52832_FEATHER"
  #elif defined(NRF52840_FEATHER_SENSE)
    #define BOARD_TYPE      "NRF52840_FEATHER_SENSE"
  #elif defined(NRF52840_ITSYBITSY)
    #define BOARD_TYPE      "NRF52840_ITSYBITSY"
  #elif defined(NRF52840_CIRCUITPLAY)
    #define BOARD_TYPE      "NRF52840_CIRCUITPLAY"
  #elif defined(NRF52840_CLUE)
    #define BOARD_TYPE      "NRF52840_CLUE"
  #elif defined(NRF52840_METRO)
    #define BOARD_TYPE      "NRF52840_METRO"
  #elif defined(NRF52840_PCA10056)
    #define BOARD_TYPE      "NRF52840_PCA10056"
  #elif defined(NINA_B302_ublox)
    #define BOARD_TYPE      "NINA_B302_ublox"
  #elif defined(NINA_B112_ublox)
    #define BOARD_TYPE      "NINA_B112_ublox"
  #elif defined(PARTICLE_XENON)
    #define BOARD_TYPE      "PARTICLE_XENON"
  #elif defined(ARDUINO_NRF52_ADAFRUIT)
    #define BOARD_TYPE      "ARDUINO_NRF52_ADAFRUIT"
  #else
    #define BOARD_TYPE      "nRF52 Unknown"
  #endif

#elif ( defined(CORE_TEENSY) )
  // Must define these pins for Teensy
  #if !defined(SerialNB)
    #warning Using default SerialNB = Serial1

    #define SerialNB   Serial1
  #else
    #warning Using SerialNB from sketch  
  #endif
  
  #if !( defined(NB_RESETN) && defined(NB_PWR) )
    #warning Using default NB_RESETN(10), NB_PWR(11)
    #define NB_RESETN  (10u)
    #define NB_PWR     (11u)

  #else
    #warning Using NB_RESETN and NB_PWR pins from sketch  
  #endif
  
  #warning Using Teensy
  //#error Teensy not supported yet
  
  #if defined(__IMXRT1062__)
    // For Teensy 4.1/4.0
    #if defined(ARDUINO_TEENSY41)
      #define BOARD_TYPE      "TEENSY 4.1"
      // Use true for NativeEthernet Library, false if using other Ethernet libraries
      #define USE_NATIVE_ETHERNET     true
    #elif defined(ARDUINO_TEENSY40)
      #define BOARD_TYPE      "TEENSY 4.0"
    #else
      #define BOARD_TYPE      "TEENSY 4.x"
    #endif      
  #elif defined(__MK66FX1M0__)
    #define BOARD_TYPE "Teensy 3.6"
  #elif defined(__MK64FX512__)
    #define BOARD_TYPE "Teensy 3.5"
  #elif defined(__MKL26Z64__)
    #define BOARD_TYPE "Teensy LC"
  #elif defined(__MK20DX256__)
    #define BOARD_TYPE "Teensy 3.2" // and Teensy 3.1 (obsolete)
  #elif defined(__MK20DX128__)
    #define BOARD_TYPE "Teensy 3.0"
  #elif defined(__AVR_AT90USB1286__)
    #error Teensy 2.0++ not supported yet
  #elif defined(__AVR_ATmega32U4__)
    #error Teensy 2.0 not supported yet
  #else
    // For Other Boards
    #define BOARD_TYPE      "Unknown Teensy Board"
  #endif

#elif ( defined(ESP8266) )

  // Must define these pins for ESP8266
  #if !defined(SerialNB)
    #warning Using default SerialNB = Serial => can not use Serial for Debug Terminal

    #define SerialNB   Serial
  #else
    #warning Using SerialNB from sketch  
  #endif
  
  #if !( defined(NB_RESETN) && defined(NB_PWR) )
    #warning Using default NB_RESETN(10), NB_PWR(11)
    #define NB_RESETN  (10u)
    #define NB_PWR     (11u)

  #else
    #warning Using NB_RESETN and NB_PWR pins from sketch  
  #endif

  #warning Using ESP8266
  
  // For ESP8266
  #warning Use ESP8266 architecture
  #define NB_USE_ESP8266
  #define BOARD_TYPE      ARDUINO_BOARD

#elif ( defined(ESP32) )

  // Must define these pins for ESP32
  #if !defined(SerialNB)
    #warning Using default SerialNB = Serial2

    #define SerialNB   Serial2
  #else
    #warning Using SerialNB from sketch  
  #endif
  
  #if !( defined(NB_RESETN) && defined(NB_PWR) )
    #warning Using default NB_RESETN(10), NB_PWR(11)
    #define NB_RESETN  (10u)
    #define NB_PWR     (11u)

  #else
    #warning Using NB_RESETN and NB_PWR pins from sketch  
  #endif

  #warning Using ESP32
  
  // For ESP32
  #warning Use ESP32 architecture
  #define NB_USE_ESP32
  #define BOARD_TYPE      ARDUINO_BOARD
  
#elif (NB_USE_STM32)

  // Must define these pins for STM32
  #if !defined(SerialNB)
    #warning Using default SerialNB = Serial1

    #define SerialNB   Serial1
  #else
    #warning Using SerialNB from sketch  
  #endif
  
  #if !( defined(NB_RESETN) && defined(NB_PWR) )
    #warning Using default NB_RESETN(10), NB_PWR(11)
    #define NB_RESETN  (10u)
    #define NB_PWR     (11u)

  #else
    #warning Using NB_RESETN and NB_PWR pins from sketch  
  #endif
 
  #warning SerialNB using SERIAL_PORT_HARDWARE, can be Serial or Serial1. See your board variant.h
  #define SerialNB     SERIAL_PORT_HARDWARE

  #if defined(STM32F0)
    #warning STM32F0 board selected
    #define BOARD_TYPE  "STM32F0"
  #elif defined(STM32F1)
    #warning STM32F1 board selected
    #define BOARD_TYPE  "STM32F1"
  #elif defined(STM32F2)
    #warning STM32F2 board selected
    #define BOARD_TYPE  "STM32F2"
  #elif defined(STM32F3)
    #warning STM32F3 board selected
    #define BOARD_TYPE  "STM32F3"
  #elif defined(STM32F4)
    #warning STM32F4 board selected
    #define BOARD_TYPE  "STM32F4"
  #elif defined(STM32F7)
  
    #if defined(ARDUINO_NUCLEO_F767ZI)
      #warning Nucleo-144 NUCLEO_F767ZI board selected, using HardwareSerial Serial1 @ pin D0/RX and D1/TX
      #define BOARD_TYPE  "NUCLEO_F767ZI"
      // RX TX
      HardwareSerial Serial1(D0, D1);
    #else
    
      #warning STM32F7 board selected
      #define BOARD_TYPE  "STM32F7"

    #endif
   
  #elif defined(STM32L0)
  
    #if defined(ARDUINO_NUCLEO_L053R8)
      #warning Nucleo-64 NUCLEO_L053R8 board selected, using HardwareSerial Serial1 @ pin D0/RX and D1/TX

    #else
    
      #warning STM32L0 board selected
      #define BOARD_TYPE  "STM32L0"

    #endif
    
  #elif defined(STM32L1)
    #warning STM32L1 board selected
    #define BOARD_TYPE  "STM32L1"
  #elif defined(STM32L4)
    #warning STM32L4 board selected
    #define BOARD_TYPE  "STM32L4"
  #elif defined(STM32H7)
    #warning STM32H7 board selected
    #define BOARD_TYPE  "STM32H7"
  #elif defined(STM32G0)
    #warning STM32G0 board selected
    #define BOARD_TYPE  "STM32G0"
  #elif defined(STM32G4)
    #warning STM32G4 board selected
    #define BOARD_TYPE  "STM32G4"
  #elif defined(STM32WB)
    #warning STM32WB board selected
    #define BOARD_TYPE  "STM32WB"
  #elif defined(STM32MP1)
    #warning STM32MP1 board selected
    #define BOARD_TYPE  "STM32MP1"
  #else
    #warning STM32 unknown board selected
    #define BOARD_TYPE  "STM32 Unknown"
  #endif  

#elif (NB_USE_RP2040)

  // Must define these pins for RP2040
  #if !defined(SerialNB)
    #warning Using default SerialNB = Serial1

    #define SerialNB   Serial1
  #else
    #warning Using SerialNB from sketch  
  #endif
  
  #if !( defined(NB_RESETN) && defined(NB_DTR) )
    #warning Using default NB_RESETN(10), NB_DTR(11)
    #define NB_RESETN  (10u)
    #define NB_DTR     (11u)

  #else
    #warning Using NB_RESETN and NB_DTR pins from sketch  
  #endif
   
  #if defined(ARDUINO_ARCH_MBED)

    #warning Using RP2040 and ARDUINO_ARCH_MBED
  
    #if ( defined(ARDUINO_NANO_RP2040_CONNECT)    || defined(ARDUINO_RASPBERRY_PI_PICO) || \
          defined(ARDUINO_GENERIC_RP2040) || defined(ARDUINO_ADAFRUIT_FEATHER_RP2040) )
      // Only undef known BOARD_NAME to use better one
      #undef BOARD_NAME
    #endif
    
    #if defined(ARDUINO_RASPBERRY_PI_PICO)
      #define BOARD_NAME      "MBED RASPBERRY_PI_PICO"
    #elif defined(ARDUINO_ADAFRUIT_FEATHER_RP2040)
      #define BOARD_NAME      "MBED ADAFRUIT_FEATHER_RP2040"
    #elif defined(ARDUINO_GENERIC_RP2040)
      #define BOARD_NAME      "MBED GENERIC_RP2040"
    #elif defined(ARDUINO_NANO_RP2040_CONNECT) 
      #define BOARD_NAME      "MBED NANO_RP2040_CONNECT"
    #else
      // Use default BOARD_NAME if exists
      #if !defined(BOARD_NAME)
        #define BOARD_NAME      "MBED Unknown RP2040"
      #endif
    #endif
   
  #else
  
      #warning Using RP2040 and arduino-pico core
        
  #endif  
  
#else

  // Must define these pins for this board
  #if !defined(SerialNB)
    #warning Using default SerialNB = Serial1

    #define SerialNB   Serial1
  #else
    #warning Using SerialNB from sketch  
  #endif
  
  #if !( defined(NB_RESETN) && defined(NB_PWR) )
    #warning Using default NB_RESETN(10), NB_PWR(11)
    #define NB_RESETN  (10u)
    #define NB_PWR     (11u)

  #else
    #warning Using NB_RESETN and NB_PWR pins from sketch  
  #endif

  //#error Using Unsupported Board
 
  #define BOARD_TYPE            "Unknown"
#endif

#ifndef BOARD_NAME
  #define BOARD_NAME    BOARD_TYPE
#endif


#endif    // _NB_BOARDS_GENERIC_H_INCLUDED
